File: phases/12-cryptography-and-security/11-public-key-iii-elliptic-curves-ed25519/code/main.py
"""
Public Key III — Elliptic Curves & Ed25519
Phase 12 — Cryptography & Security

Elliptic curve point operations and Ed25519 key operations.
"""


class EllipticCurve:
    """An elliptic curve y^2 = x^3 + ax + b over a prime field F_p."""

    def __init__(self, a, b, p):
        self.a = a
        self.b = b
        self.p = p
        d = (4 * a**3 + 27 * b**2) % p
        if d == 0:
            raise ValueError(f"singular curve: discriminant is zero (a={a}, b={b}, p={p})")

    def is_on_curve(self, x, y):
        """Check if (x, y) satisfies y^2 ≡ x^3 + ax + b (mod p)."""
        return (y * y - (x * x * x + self.a * x + self.b)) % self.p == 0


class Point:
    """A point on an elliptic curve. The point at infinity is (None, None)."""

    def __init__(self, curve, x, y):
        self.curve = curve
        self.x = x
        self.y = y
        if x is not None and y is not None:
            if not curve.is_on_curve(x, y):
                raise ValueError(f"({x}, {y}) is not on the curve")
            self.x = x % curve.p
            self.y = y % curve.p

    def __eq__(self, other):
        if not isinstance(other, Point):
            return NotImplemented
        return self.x == other.x and self.y == other.y and self.curve == other.curve

    def __ne__(self, other):
        return not self.__eq__(other)

    def __repr__(self):
        if self.is_infinity():
            return "Point(infinity)"
        return f"Point({self.x}, {self.y})"

    def is_infinity(self):
        return self.x is None and self.y is None

    def negate(self):
        """Return the additive inverse of this point."""
        if self.is_infinity():
            return self
        return Point(self.curve, self.x, (-self.y) % self.curve.p)


def extended_gcd(a, b):
    """Extended Euclidean algorithm: returns (gcd, x, y) such that ax + by = gcd."""
    if a == 0:
        return b, 0, 1
    g, x1, y1 = extended_gcd(b % a, a)
    return g, y1 - (b // a) * x1, x1


def inv_mod(k, p):
    """Modular inverse of k modulo p (p must be prime)."""
    if k == 0:
        raise ZeroDivisionError("cannot invert zero")
    g, x, _ = extended_gcd(k % p, p)
    if g != 1:
        raise ValueError(f"{k} is not invertible mod {p}")
    return x % p


def point_add(P, Q):
    """Add two points on the same elliptic curve."""
    if P.is_infinity():
        return Q
    if Q.is_infinity():
        return P

    assert P.curve is Q.curve, "points must be on the same curve"

    curve = P.curve

    # P + (-P) = infinity
    if P.x == Q.x and (P.y + Q.y) % curve.p == 0:
        return Point(curve, None, None)

    # Compute slope
    if P.x == Q.x:
        # Point doubling: tangent slope = (3x^2 + a) / (2y)
        s_num = (3 * P.x * P.x + curve.a) % curve.p
        s_den = (2 * P.y) % curve.p
    else:
        # Point addition: secant slope = (y2 - y1) / (x2 - x1)
        s_num = (Q.y - P.y) % curve.p
        s_den = (Q.x - P.x) % curve.p

    s = s_num * inv_mod(s_den, curve.p) % curve.p

    x3 = (s * s - P.x - Q.x) % curve.p
    y3 = (s * (P.x - x3) - P.y) % curve.p
    return Point(curve, x3, y3)


def scalar_mult(k, P):
    """Multiply point P by scalar k using double-and-add."""
    if k == 0 or P.is_infinity():
        return Point(P.curve, None, None)

    result = Point(P.curve, None, None)
    addend = P

    n = k if k >= 0 else -k
    while n:
        if n & 1:
            result = point_add(result, addend)
        addend = point_add(addend, addend)
        n >>= 1

    if k < 0:
        result = result.negate()

    return result


def find_order(G, max_order=1000):
    """Find the order of point G by brute force multiplication."""
    if G.is_infinity():
        return 1
    cur = G
    for i in range(1, max_order + 1):
        cur = point_add(cur, G)
        if cur.is_infinity():
            return i + 1
    return None


def count_points_brute(curve):
    """Count all points on the curve by iterating over all x in F_p (only for small p)."""
    count = 1  # point at infinity
    for x in range(curve.p):
        rhs = (x * x * x + curve.a * x + curve.b) % curve.p
        # Check if rhs is a quadratic residue (Euler's criterion for small fields)
        for y in range(curve.p):
            if (y * y) % curve.p == rhs:
                count += 1
    return count


def demo_ecc():
    """Demonstrate elliptic curve point operations on a small curve."""
    print("=== Elliptic Curve Operations Demo ===\n")

    # Small curve: y^2 = x^3 + 2x + 3 mod 97
    curve = EllipticCurve(a=2, b=3, p=97)
    G = Point(curve, 3, 6)

    print(f"Curve: y^2 = x^3 + {curve.a}x + {curve.b} mod {curve.p}")
    print(f"Generator G = {G}")
    print(f"G on curve: {curve.is_on_curve(G.x, G.y)}")
    print()

    # Point addition: G + G
    G2 = point_add(G, G)
    print(f"2*G = {G2}")
    assert curve.is_on_curve(G2.x, G2.y)

    # Verify from problem statement: 2*(3,6) = (80,10)
    assert G2.x == 80 and G2.y == 10, f"expected (80,10), got ({G2.x},{G2.y})"
    print("  ✓ Verified: 2*(3,6) = (80,10)")
    print()

    # Scalar multiplication
    G3 = scalar_mult(3, G)
    G4 = scalar_mult(4, G)
    G5 = scalar_mult(5, G)
    print(f"3*G = {G3}")
    print(f"4*G = {G4}")
    print(f"5*G = {G5}")
    print()

    # Verify order: 5*G should be infinity (generator order is 5)
    assert G5.is_infinity(), f"expected infinity, got {G5}"
    print("  ✓ Verified: 5*G = infinity (order of G is 5)")
    print()

    # Associativity: (G + G) + G == G + (G + G)
    left = point_add(G2, G)
    right = point_add(G, G2)
    assert left == right
    print(f"  ✓ Associativity: (G+G)+G == G+(G+G) == {left}")
    print()

    # Commutativity: G + 2G == 2G + G (already tested above)
    assert left == right
    print(f"  ✓ Commutativity: G+2G == 2G+G == {left}")
    print()

    # Negation: P + (-P) = infinity
    neg_G = G.negate()
    inf = point_add(G, neg_G)
    assert inf.is_infinity()
    print(f"  ✓ Negation: G + (-G) = infinity")
    print(f"      G  = {G}")
    print(f"     -G  = {neg_G}")
    print()

    # Scalar multiplication consistency: 2*(5*G) == 5*(2*G) == infinity
    G10a = scalar_mult(2, G5)   # 2 * infinity = infinity
    G10b = scalar_mult(5, G2)   # 5 * (2*G) = 10*G = infinity
    assert G10a.is_infinity()
    assert G10b.is_infinity()
    print(f"  ✓ Consistency: 2*(5*G) == 5*(2*G) == infinity")
    print()

    # Negative scalar multiplication
    neg_P = scalar_mult(-3, G)
    check = point_add(G3, neg_P)
    assert check.is_infinity(), f"negative scalar failed: {check}"
    print(f"  ✓ Negative scalar: 3*G + (-3)*G = infinity")
    print()

    # Find the order of the generator
    order = find_order(G)
    print(f"  Order of G: {order}")
    print()

    # Count total points (Hasse bound: |#E - (p+1)| ≤ 2*sqrt(p))
    total = count_points_brute(curve)
    hasse_lower = curve.p + 1 - int(2 * (curve.p ** 0.5))
    hasse_upper = curve.p + 1 + int(2 * (curve.p ** 0.5))
    print(f"  Total points on curve: {total}")
    print(f"  Hasse bound: [{hasse_lower}, {hasse_upper}]")
    assert hasse_lower <= total <= hasse_upper, "Hasse bound violated!"
    print(f"  ✓ Within Hasse bound")
    print()

    # ECDH-style key exchange
    print("=== ECDH-style Key Exchange ===")
    alice_priv = 17
    bob_priv = 23
    alice_pub = scalar_mult(alice_priv, G)
    bob_pub = scalar_mult(bob_priv, G)

    # Handle case where public key is infinity (shouldn't happen with valid privates)
    assert not alice_pub.is_infinity()
    assert not bob_pub.is_infinity()

    shared_alice = scalar_mult(alice_priv, bob_pub)
    shared_bob = scalar_mult(bob_priv, alice_pub)
    shared_match = shared_alice == shared_bob

    print(f"  Alice's private: {alice_priv}")
    print(f"  Bob's private:   {bob_priv}")
    print(f"  Alice's public:  {alice_pub}")
    print(f"  Bob's public:    {bob_pub}")
    print(f"  Alice's shared:  {shared_alice}")
    print(f"  Bob's shared:    {shared_bob}")
    print(f"  Shared secrets match: {shared_match}")
    assert shared_match
    print()

    # Edge case: P + infinity = P, infinity + P = P
    inf = Point(curve, None, None)
    assert point_add(G, inf) == G
    assert point_add(inf, G) == G
    assert point_add(inf, inf).is_infinity()
    print(f"  ✓ Identity: P + O = P, O + P = P, O + O = O")
    print()

    # Edge case: scalar multiplication by zero
    zeroP = scalar_mult(0, G)
    assert zeroP.is_infinity()
    print(f"  ✓ Scalar mult by 0: 0*G = infinity")
    print()

    # Another curve for variety: y^2 = x^3 + 7 mod 71 (like secp256k1 but tiny)
    print("=== Second Curve: y^2 = x^3 + 7 mod 71 ===")
    curve2 = EllipticCurve(a=0, b=7, p=71)
    G2 = Point(curve2, 1, 24)  # Known point on this curve: 24^2 ≡ 1^3 + 7 (mod 71)
    print(f"  Generator: {G2}")
    print(f"  2*G = {point_add(G2, G2)}")
    print(f"  3*G = {scalar_mult(3, G2)}")
    order2 = find_order(G2)
    print(f"  Order of G2: {order2}")
    print()

    # Key size comparison
    print("=== Key Size Comparison ===")
    print(f"  ECC-256:  32 bytes  (equivalent security to RSA-3072)")
    print(f"  ECC-384:  48 bytes  (equivalent security to RSA-7680)")
    print(f"  ECC-521:  66 bytes  (equivalent security to RSA-15360)")
    print(f"  RSA-2048: 256 bytes")
    print(f"  RSA-4096: 512 bytes")
    print()

    print("=== Ed25519 Overview ===")
    print(f"  Ed25519 key generation: seed(32) + SHA-512 + clamp + scalar_mult(B, k)")
    print(f"  Ed25519 signing: deterministic, no RNG needed")
    print(f"  Ed25519 verification: double-base scalar multiplication")
    print(f"  Benefits: fast, secure, constant-time, no RNG dependency")
    print(f"  Key size: 32-byte secret + 32-byte public = 64 bytes total")
    print(f"  Signature size: 64 bytes (vs RSA-2048's 256 bytes)")
    print()

    print("All tests passed!")


def main():
    demo_ecc()


if __name__ == "__main__":
    main()

File: phases/12-cryptography-and-security/11-public-key-iii-elliptic-curves-ed25519/code/main.rs
//! Public Key III — Elliptic Curves & Ed25519
//! Phase 12 — Cryptography & Security
//!
//! Demonstrates elliptic curve point operations and Ed25519 key operations.

/// An elliptic curve y^2 = x^3 + ax + b over F_p.
struct EllipticCurve {
    a: i64,
    b: i64,
    p: i64,
}

/// A point on an elliptic curve. `(None, None)` represents the point at infinity.
#[derive(Debug, Clone, Copy, PartialEq)]
struct Point {
    x: Option<i64>,
    y: Option<i64>,
}

const INF: Point = Point {
    x: None,
    y: None,
};

impl EllipticCurve {
    fn new(a: i64, b: i64, p: i64) -> Self {
        let disc = ((4 * a * a * a + 27 * b * b) % p + p) % p;
        assert!(disc != 0, "singular curve: discriminant is zero");
        EllipticCurve { a, b, p }
    }

    fn is_on_curve(&self, x: i64, y: i64) -> bool {
        let lhs = ((y * y) % self.p + self.p) % self.p;
        let rhs = ((x * x * x + self.a * x + self.b) % self.p + self.p) % self.p;
        lhs == rhs
    }
}

impl Point {
    fn new(curve: &EllipticCurve, x: Option<i64>, y: Option<i64>) -> Self {
        if let (Some(xv), Some(yv)) = (x, y) {
            assert!(curve.is_on_curve(xv, yv), "point ({}, {}) is not on the curve", xv, yv);
        }
        Point { x, y }
    }

    fn infinity() -> Self {
        INF
    }

    fn is_infinity(&self) -> bool {
        self.x.is_none() && self.y.is_none()
    }

    fn negate(&self, curve: &EllipticCurve) -> Self {
        if self.is_infinity() {
            return *self;
        }
        let x = self.x.unwrap();
        let y = self.y.unwrap();
        Point::new(curve, Some(x), Some((-(y % curve.p) + curve.p) % curve.p))
    }
}

/// Extended Euclidean algorithm: returns (gcd, x, y) such that ax + by = gcd.
fn extended_gcd(a: i64, b: i64) -> (i64, i64, i64) {
    if a == 0 {
        return (b, 0, 1);
    }
    let (g, x1, y1) = extended_gcd(b % a, a);
    (g, y1 - (b / a) * x1, x1)
}

/// Modular inverse of k modulo m.
fn inv_mod(k: i64, m: i64) -> i64 {
    assert_ne!(k, 0, "cannot invert zero");
    let k = (k % m + m) % m;
    let (g, x, _) = extended_gcd(k, m);
    assert_eq!(g, 1, "{} is not invertible mod {}", k, m);
    (x % m + m) % m
}

/// Add two points on the same elliptic curve.
fn point_add(P: &Point, Q: &Point, curve: &EllipticCurve) -> Point {
    if P.is_infinity() {
        return *Q;
    }
    if Q.is_infinity() {
        return *P;
    }

    let px = P.x.unwrap();
    let py = P.y.unwrap();
    let qx = Q.x.unwrap();
    let qy = Q.y.unwrap();

    // P + (-P) = infinity
    if px == qx && (py + qy) % curve.p == 0 {
        return Point::infinity();
    }

    let s: i64;
    if px == qx {
        // Point doubling: tangent slope = (3x^2 + a) / (2y)
        let num = ((3 * px * px + curve.a) % curve.p + curve.p) % curve.p;
        let den = (2 * py) % curve.p;
        s = num * inv_mod(den, curve.p) % curve.p;
    } else {
        // Point addition: secant slope = (y2 - y1) / (x2 - x1)
        let num = ((qy - py) % curve.p + curve.p) % curve.p;
        let den = ((qx - px) % curve.p + curve.p) % curve.p;
        s = num * inv_mod(den, curve.p) % curve.p;
    }

    let x3 = ((s * s - px - qx) % curve.p + curve.p) % curve.p;
    let y3 = ((s * (px - x3) - py) % curve.p + curve.p) % curve.p;

    Point::new(curve, Some(x3), Some(y3))
}

/// Multiply point P by scalar k using double-and-add.
fn scalar_mult(k: i64, P: &Point, curve: &EllipticCurve) -> Point {
    if k == 0 || P.is_infinity() {
        return Point::infinity();
    }

    let mut result = Point::infinity();
    let mut addend = *P;
    let mut n = k.abs();

    while n > 0 {
        if n & 1 == 1 {
            result = point_add(&result, &addend, curve);
        }
        addend = point_add(&addend, &addend, curve);
        n >>= 1;
    }

    if k < 0 {
        result = result.negate(curve);
    }

    result
}

/// Find the order of a point by brute-force scalar multiplication (only for small groups).
fn find_order(G: &Point, curve: &EllipticCurve, max_order: i64) -> Option<i64> {
    if G.is_infinity() {
        return Some(1);
    }
    let mut cur = *G;
    for i in 1..=max_order {
        cur = point_add(&cur, G, curve);
        if cur.is_infinity() {
            return Some(i + 1);
        }
    }
    None
}

/// Count all points on the curve by brute force (only for small p).
fn count_points_brute(curve: &EllipticCurve) -> i64 {
    let mut count = 1; // point at infinity
    for x in 0..curve.p {
        let rhs = ((x * x * x + curve.a * x + curve.b) % curve.p + curve.p) % curve.p;
        for y in 0..curve.p {
            if (y * y) % curve.p == rhs {
                count += 1;
            }
        }
    }
    count
}

fn main() {
    println!("=== Elliptic Curve Operations Demo ===\n");

    // Small curve: y^2 = x^3 + 2x + 3 mod 97
    let curve = EllipticCurve::new(2, 3, 97);
    let G = Point::new(&curve, Some(3), Some(6));

    println!("Curve: y^2 = x^3 + {}x + {} mod {}", curve.a, curve.b, curve.p);
    println!("Generator G = (3, 6)");
    println!("G on curve: {}", curve.is_on_curve(3, 6));
    println!();

    // 2*G
    let G2 = point_add(&G, &G, &curve);
    println!("2*G = ({}, {})", G2.x.unwrap(), G2.y.unwrap());
    assert!(curve.is_on_curve(G2.x.unwrap(), G2.y.unwrap()));
    assert_eq!(G2.x.unwrap(), 80);
    assert_eq!(G2.y.unwrap(), 10);
    println!("  ✓ Verified: 2*(3,6) = (80,10)");
    println!();

    // Scalar multiplication
    let G3 = scalar_mult(3, &G, &curve);
    let G4 = scalar_mult(4, &G, &curve);
    let G5 = scalar_mult(5, &G, &curve);
    println!("3*G = ({}, {})", G3.x.unwrap(), G3.y.unwrap());
    println!("4*G = ({}, {})", G4.x.unwrap(), G4.y.unwrap());
    println!("5*G = ({:?}, {:?})", G5.x, G5.y);
    println!();

    // Verify generator order is 5
    assert!(G5.is_infinity(), "order of G should be 5");
    println!("  ✓ Verified: 5*G = infinity (order of G is 5)");
    println!();

    // Associativity: (G+G)+G == G+(G+G)
    let left = point_add(&G2, &G, &curve);
    let right = point_add(&G, &G2, &curve);
    assert_eq!(left, right);
    println!("  ✓ Associativity: (G+G)+G == G+(G+G) == ({}, {})",
             left.x.unwrap(), left.y.unwrap());
    println!();

    // Negation: P + (-P) = infinity
    let neg_G = G.negate(&curve);
    let inf = point_add(&G, &neg_G, &curve);
    assert!(inf.is_infinity());
    println!("  ✓ Negation: G + (-G) = infinity");
    println!("      G  = ({}, {})", G.x.unwrap(), G.y.unwrap());
    println!("     -G  = ({}, {})", neg_G.x.unwrap(), neg_G.y.unwrap());
    println!();

    // Scalar multiplication consistency
    let G10a = scalar_mult(2, &G5, &curve);
    let G10b = scalar_mult(5, &G2, &curve);
    assert!(G10a.is_infinity());
    assert!(G10b.is_infinity());
    println!("  ✓ Consistency: 2*(5*G) == 5*(2*G) == infinity");
    println!();

    // Negative scalar multiplication
    let neg3 = scalar_mult(-3, &G, &curve);
    let check = point_add(&G3, &neg3, &curve);
    assert!(check.is_infinity());
    println!("  ✓ Negative scalar: 3*G + (-3)*G = infinity");
    println!();

    // Find order of the generator
    match find_order(&G, &curve, 100) {
        Some(order) => println!("  Order of G: {}", order),
        None => println!("  Order not found within limit"),
    }
    println!();

    // Count total points (Hasse bound)
    let total = count_points_brute(&curve);
    let hasse_lower = curve.p + 1 - (2.0 * (curve.p as f64).sqrt()) as i64;
    let hasse_upper = curve.p + 1 + (2.0 * (curve.p as f64).sqrt()) as i64;
    println!("  Total points on curve: {}", total);
    println!("  Hasse bound: [{}, {}]", hasse_lower, hasse_upper);
    assert!(total >= hasse_lower && total <= hasse_upper, "Hasse bound violated!");
    println!("  ✓ Within Hasse bound");
    println!();

    // ECDH-style key exchange
    println!("=== ECDH-style Key Exchange ===");
    let alice_priv: i64 = 17;
    let bob_priv: i64 = 23;
    let alice_pub = scalar_mult(alice_priv, &G, &curve);
    let bob_pub = scalar_mult(bob_priv, &G, &curve);

    assert!(!alice_pub.is_infinity());
    assert!(!bob_pub.is_infinity());

    let shared_alice = scalar_mult(alice_priv, &bob_pub, &curve);
    let shared_bob = scalar_mult(bob_priv, &alice_pub, &curve);
    let shared_match = shared_alice == shared_bob;

    println!("  Alice's private: {}", alice_priv);
    println!("  Bob's private:   {}", bob_priv);
    println!("  Alice's public:  ({}, {})", alice_pub.x.unwrap(), alice_pub.y.unwrap());
    println!("  Bob's public:    ({}, {})", bob_pub.x.unwrap(), bob_pub.y.unwrap());
    println!("  Shared match: {}", shared_match);
    assert!(shared_match);
    println!();

    // Edge case: identity operations
    let inf_pt = Point::infinity();
    assert_eq!(point_add(&G, &inf_pt, &curve), G);
    assert_eq!(point_add(&inf_pt, &G, &curve), G);
    assert!(point_add(&inf_pt, &inf_pt, &curve).is_infinity());
    println!("  ✓ Identity: P+O = P, O+P = P, O+O = O");
    println!();

    // Edge case: scalar multiplication by zero
    let zero = scalar_mult(0, &G, &curve);
    assert!(zero.is_infinity());
    println!("  ✓ Scalar mult by 0: 0*G = infinity");
    println!();

    // Edge case: scalar multiplication by 1
    let one = scalar_mult(1, &G, &curve);
    assert_eq!(one, G);
    println!("  ✓ Scalar mult by 1: 1*G = G");
    println!();

    // Second curve: y^2 = x^3 + 7 mod 71 (like secp256k1 but tiny)
    println!("=== Second Curve: y^2 = x^3 + 7 mod 71 ===");
    let curve2 = EllipticCurve::new(0, 7, 71);
    let G2 = Point::new(&curve2, Some(1), Some(24));
    println!("  Generator: (1, 13)");
    let G2_2 = point_add(&G2, &G2, &curve2);
    println!("  2*G = ({}, {})", G2_2.x.unwrap(), G2_2.y.unwrap());
    let G2_3 = scalar_mult(3, &G2, &curve2);
    println!("  3*G = ({}, {})", G2_3.x.unwrap(), G2_3.y.unwrap());
    match find_order(&G2, &curve2, 200) {
        Some(o) => println!("  Order of G2: {}", o),
        None => println!("  Order not found within limit"),
    }
    println!();

    // Key size comparison
    println!("=== Key Size Comparison ===");
    println!("  ECC-256:  32 bytes  (equivalent security to RSA-3072)");
    println!("  ECC-384:  48 bytes  (equivalent security to RSA-7680)");
    println!("  ECC-521:  66 bytes  (equivalent security to RSA-15360)");
    println!("  RSA-2048: 256 bytes");
    println!("  RSA-4096: 512 bytes");
    println!();

    println!("=== Ed25519 Overview ===");
    println!("  Ed25519 key generation: seed(32) + SHA-512 + clamp + scalar_mult(B, k)");
    println!("  Ed25519 signing: deterministic, no RNG needed");
    println!("  Ed25519 verification: double-base scalar multiplication");
    println!("  Benefits: fast, secure, constant-time, no RNG dependency");
    println!("  Key size: 32-byte secret + 32-byte public = 64 bytes total");
    println!("  Signature size: 64 bytes (vs RSA-2048's 256 bytes)");
    println!();

    println!("All tests passed!");
}
